### PR TITLE
Import Gujarati

### LINKED ISSUE #525: Add Gujarati script
Sources will be on the repository Sept 16

cc @Hoolean @madig @belluzj

### PR DESCRIPTION
**Pull request by Hoolean**
Hello all! This PR tracks #525.

We have rebased on `import-sinhala`, although if Gujarati becomes ready before Sinhala, we can rebase on `main` instead.

Before the branch is ready for review, we must check for any CI warnings or errors. In addition, we will add the shaping tests from @kalapi when they are delivered : )

---

_Closes #525._

### CODE REVIEW

**Comment by Hoolean** [comment]
Hello @kalapi!

Thank you for providing the shaping tests : )

Having updated the glyph sets, it appears that the following glyphs are unreachable:

- /iMatra_anusvara-gujarati.alt01
- /iMatra_reph-gujarati.alt01
- /iMatra_reph_anusvara-gujarati.alt01
- /t_r-gujarati

Do they require a codepoint or feature to make them accessible, or are they not intended to be used?

**Comment by MariannaPaszkowska** [comment]
@kalapi  Could you please check mark positioning here:
Combining outline issue - design or positioning of component/anchor:
ખુ ખૂ ઘૃ ખ્રૃ ખૄ ઘૄ ગ્રૄ ઙ્ક્ષ ઙ્ઘ ઙ્મ ટ્વ ડ્ન ડ્વ ઢ્ન 
 
Screenshot 2022-09-30 at 2 42 39 PM

_Screenshot from Figma_

Could you also confirm if this is intentional:

Component positioning consistency Text vs Regular: 
દ્ગુ

Screenshot 2022-09-30 at 2 09 51 PM

_Screenshot from Figma_

cc @chrissimpkins

**Comment by kalapi** [comment]
@MariannaPaszkowska This is strange, checking now. Thanks for spotting it.

**Comment by kalapi** [comment]
@MariannaPaszkowska I think this is a Figma issue. Here's a screenshot from Chrome:

Display
Screenshot 2022-09-30 at 12 34 56 PM
Text
Screenshot 2022-09-30 at 12 36 58 PM

I see that the 4th combination above doesn't position correctly but that's an impossible combination, so it's fine. But the others are shaping fine. 

The 2nd issue (દ્ગુ uMatra positioning) is also not a problem on Chrome (Harfbuzz). How do we fix this?

**Comment by kalapi** [comment]
> Having updated the glyph sets, it appears that the following glyphs are unreachable:
> 
> * /iMatra_anusvara-gujarati.alt01
> * /iMatra_reph-gujarati.alt01
> * /iMatra_reph_anusvara-gujarati.alt01
> * /t_r-gujarati

@Hoolean The attached package fixes the above. 

[2022-09-30_GoogleSansGujarati_V1.000_BETA03.zip](https://github.com/googlefonts/googlesans/files/9686449/2022-09-30_GoogleSansGujarati_V1.000_BETA03.zip)

**Comment by MariannaPaszkowska** [comment]
Thanks for the update, @kalapi!
I can confirm that this is not a problem on Chrome (Harfbuzz).

Screenshot 2022-10-03 at 9 21 38 AM

_Screenshot from Wakamaifondue._



There is an improvement in Figma now, shaping is different, but now other issues have popped up: 
Screenshot 2022-10-03 at 9 14 32 AM

_Screenshot from Figma._

@Hoolean @belluzj Any ideas on what might be causing this and how can we fix it?

**Comment by belluzj** [comment]
Since you have a special relationship with Figma, would you be able to ask them what version of Harfbuzz they're compiling against in their C++/WASM bundle? They may want to update it if it's old? Just a thought

**Comment by kalapi** [comment]
@MariannaPaszkowska I can confirm that I haven't touched any of the positioning in the new Gujarati package other than in a few conjuncts not listed above.

**Comment by MariannaPaszkowska** [comment]
@kalapi I also suspect that it is something Figma-specific. @Hoolean @belluzj We can continue with the production and release. 
I tested GS Gujarati in other environments and can confirm that except for Figma, I see expected behavior:

Screenshot 2022-10-04 at 9 50 25 AM

_TextEdit on Apple_  *it was the same for _wakamaifondue and Samsa in Safari.

Screenshot 2022-10-04 at 9 43 40 AM

_wakamaifondue in Firefox_

cc @chrissimpkins

**Comment by madig** [comment]
No new GDEF warnings in the VFs compared to Odia.

**Comment by madig** [comment]
@kalapi can you please test the latest import in Adobe apps again?

**Comment by madig** [comment]
I believe this is ready for inspection, given that we're ignoring Figma and InDesign seems to work fine?

**Comment by madig** [comment]
@kalapi can you please reupload the beta3 package with your built fonts? I think they're from beta2?

**Comment by kalapi** [comment]
Hi @madig let me check and get back to you immediately.

**Comment by kalapi** [comment]
I checked the fonts and they seem to be generated after I made the changes but i've generated them again in the below package:

[2022-10-05_GoogleSansGujarati_V1.000_BETA03b.zip](https://github.com/googlefonts/googlesans/files/9716606/2022-10-05_GoogleSansGujarati_V1.000_BETA03b.zip)

**Comment by Hoolean** [comment]
Hello @MariannaPaszkowska, @chrissimpkins,

Nikolaus has merged in the latest fixes (thank you @kalapi!), and Hanna will begin regression testing them tomorrow morning 🚀

**Comment by chrissimpkins** [comment]
> I believe this is ready for inspection, given that we're ignoring Figma and InDesign seems to work fine?

To be fair, we're not ignoring any environment.  TMK the identified issues are in product-specific font rendering environments, sometimes maybe even in a version-specific way like the issues identified in InDesign by DaMa but unable to be reproduced in InDesign by Kalapi.  Do you have recommendations about how to achieve correct shaping at the font level?  I'm not aware of an identified solution to the issue that seems to be isolated to the Figma environment in the thread here. The plan is to approach the Figma team to report this as a shaping bug in their environment since the layout code WAI in other tested environments.

**Comment by chrissimpkins** [comment]
OK to merge this before #517 if we complete the review here earlier?

**Comment by belluzj** [comment]
Hanna is scheduled to check the latest version of InDesign on our side as part of her review today.

About the merge order, any is fine, we'll fix conflicts or rebase the other one, whichever that is. Thanks!

**Comment by madig** [comment]
@kalapi please see the updated import log with the latest comments. What version of InDesign are you using? Hanna could reproduce the error in 17.3 and 17.4.

**Comment by kalapi** [comment]
Hi @madig. I have v16 and v17.4 and shaping works on both:

Screenshot 2022-10-06 at 11 38 10 AM

Screenshot 2022-10-06 at 11 46 46 AM

These are my options for 'Advanced Type':

Screenshot 2022-10-06 at 11 38 51 AM

Language selection has no effect.

Thanks for updating the import log. I'll look at the other things posted there. 

This may also be interesting although it's from 2 years ago and i'd expect Adobe to activate this by default (but maybe they don't): https://twitter.com/vinodbala/status/1285259248973213696

**Comment by kalapi** [comment]
@madig I'm attaching a new package based on Hanna's observations.

[2022-10-06_GoogleSansGujarati_V1.000_BETA04.zip](https://github.com/googlefonts/googlesans/files/9729436/2022-10-06_GoogleSansGujarati_V1.000_BETA04.zip)

**Comment by MariannaPaszkowska** [comment]
@belluzj @madig Any chance we can have this finished today?

**Comment by madig** [comment]
On it.

**Comment by madig** [comment]
So, we had the idea that InDesign doesn't seem to work the fonts out of the box because it may expect the older `gujr` shaping model instead of the newer `gujr2` one (various gujr fonts on Google Fonts like Rasa have both and work), but Kalapi only targets Harfbuzz. We good on ignoring InDesign then or requiring users to use the hidden Harfbuzz option?

**Comment by MariannaPaszkowska** [comment]
@madig Since users can use the hidden Harfbuzz option, and this environment is not our priority, let's proceed using only `gujr2`. 
CC @chrissimpkins

**Comment by madig** [comment]
I'll see if I can find anyone to cross-check the latest changes on our side.

**Comment by madig** [comment]
@kalapi Can you please give us a high-level changelog for beta 4? 1106 files are being changed...

**Comment by chrissimpkins** [comment]
I noticed that the slash.loclTAML still trails all of the new Gujarati glyphs added here. It looks like we have a merge conflict now in https://github.com/googlefonts/googlesans/pull/533.  IIUC the plan will be to merge this, then conflicts in #533 will be addressed and we will merge that?

File: scripts/one-off/check-script-tags.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+import tomli
```
Comment (chrissimpkins) [inline scripts/one-off/check-script-tags.py:15]: Please add Apache 2 license header

**Review by chrissimpkins — CHANGES_REQUESTED** [review]
Missing Apache license on the new Py script

**Comment by chrissimpkins** [comment]
Will you please push a build branch so that we can trigger the static font format checks when the import work is completed?

**Comment by Hoolean** [comment]
Hello @chrissimpkins, yes, I think that makes sense for /slash.loclTAML; we can rebase whichever branch comes second, and having that be #533 seems likes the simplest approach.

On the build, to check I understand correctly: would you like us to make the build- branch after we have cross-checked the latest fixes, or would you prefer us to set this running on the merged sources now?

**Comment by madig** [comment]
I pushed a branch now: https://github.com/googlefonts/googlesans/actions/runs/3204809760

We can update it as we go.

**Comment by kalapi** [comment]
> @kalapi Can you please give us a high-level changelog for beta 4? 1106 files are being changed...

Yes ofcourse; 

1. The majority of the changes are from the Medium-italic interpolation where a straight stem didn't have the same start point in almost all of the base glyphs resulting in the following rhomboidal slanted stems:

Screenshot 2022-10-07 at 8 56 18 AM

2. This also led me to change anchor-positions of the 'bottomright' anchor by a few units to make the connection proper.

The majority of changes are from the top two. Then,

3. Added kerning for a few glyphs

4. Added 4 new glyphs `t-gujarati.alt02`, `t_t-gujarati`, `n-gujarati.alt01`, and `l_ka-gujarati`. 

5. Updated OT layout code to support above glyphs

**Comment by chrissimpkins** [comment]
> would you like us to make the build- branch after we have cross-checked the latest fixes, or would you prefer us to set this running on the merged sources now?

Only when you feel that the changes are complete and ready for Marianna and I to review. Thanks Harry!

**Comment by madig** [comment]
Under hopefully final QA now, estimated time 90 mins (because that's when she goes home 😁).

**Comment by madig** [comment]
Our QA person couldn't finish the review, we'll try to get the original reviewer on it next week... sorry :(

She found one thing though, the interpolation fixes seem to have swapped some outlines (see comment below original interpolation issue in doc).

**Comment by kalapi** [comment]
> Our QA person couldn't finish the review, we'll try to get the original reviewer on it next week... sorry :(
> 
> She found one thing though, the interpolation fixes seem to have swapped some outlines (see comment below original interpolation issue in doc).

Thanks, will look into it.

**Comment by kalapi** [comment]
> > Our QA person couldn't finish the review, we'll try to get the original reviewer on it next week... sorry :(
> > She found one thing though, the interpolation fixes seem to have swapped some outlines (see comment below original interpolation issue in doc).
> 
> Thanks, will look into it.

Fixes minor interpolation issues in the the `opsz` axis:

Updated glyphs in the upright master:

```
ભ્ર bh_ra-gujarati
લ્ર l_ra-gujarati
```

Updated glyphs in the italic master:

```
લ la-gujarati
લ્ l-gujarati
ક્લ k_la-gujarati
ઝ્ર jh_ra-gujarati
```

[2022-10-07_GoogleSansGujarati_V1.000_BETA05.zip](https://github.com/googlefonts/googlesans/files/9736464/2022-10-07_GoogleSansGujarati_V1.000_BETA05.zip)

**Comment by Hoolean** [comment]
Hello @kalapi! Thank you for the speedy turnaround : )

We have made a few assumptions in the merge that we would like to check with you:

### 1) Feature code in upright vs italic

The feature code is different in the upright and italic fonts. We have assumed that the implementation and positioning in upright is preferred, and have used it for both. Is this correct?

##### In upright

```AFDKO
# Positioned after GUJ_dflt_pres_halfFormAlternates
lookup GUJ_dflt_pres_conjunctHalfForms {
	sub t_ta-gujarati halant-gujarati by t_t-gujarati;
	sub t-gujarati [t-gujarati t-gujarati.alt01 t-gujarati.alt02] by t_t-gujarati;
} GUJ_dflt_pres_conjunctHalfForms;
```

##### In italic

```AFDKO
# Positioned before GUJ_dflt_pres_halfFormAlternates
lookup GUJ_dflt_pres_conjunctHalfForms {
	sub t_ta-gujarati halant-gujarati by t_t-gujarati;
} GUJ_dflt_pres_conjunctHalfForms;
```

### 2) Extra glyphs on import list

A few glyphs were included on the import list that were not previously in the script; we could not see any anchor changes, and so have discarded them.

* .notdef
* NULL
* CR
* space
* nbspace
* zerowidthspace
* zerowidthjoiner
* zerowidthnonjoiner
* dottedCircle

**Comment by kalapi** [comment]
> Hello @kalapi! Thank you for the speedy turnaround : )
> 
> 
> 
> We have made a few assumptions in the merge that we would like to check with you:
> 
> 
> 
> ### 1) Feature code in upright vs italic
> 
> 
> 
> The feature code is different in the upright and italic fonts. We have assumed that the implementation and positioning in upright is preferred, and have used it for both. Is this correct?
> 
> 
> 
> ##### In upright
> 
> 
> 
> ```AFDKO
> 
> # Positioned after GUJ_dflt_pres_halfFormAlternates
> 
> lookup GUJ_dflt_pres_conjunctHalfForms {
> 
> 	sub t_ta-gujarati halant-gujarati by t_t-gujarati;
> 
> 	sub t-gujarati [t-gujarati t-gujarati.alt01 t-gujarati.alt02] by t_t-gujarati;
> 
> } GUJ_dflt_pres_conjunctHalfForms;
> 
> ```
> 
> 
> 
> ##### In italic
> 
> 
> 
> ```AFDKO
> 
> # Positioned before GUJ_dflt_pres_halfFormAlternates
> 
> lookup GUJ_dflt_pres_conjunctHalfForms {
> 
> 	sub t_ta-gujarati halant-gujarati by t_t-gujarati;
> 
> } GUJ_dflt_pres_conjunctHalfForms;
> 
> ```
> 
> 
> 
> ### 2) Extra glyphs on import list
> 
> 
> 
> A few glyphs were included on the import list that were not previously in the script; we could not see any anchor changes, and so have discarded them.
> 
> 
> 
> * .notdef
> 
> * NULL
> 
> * CR
> 
> * space
> 
> * nbspace
> 
> * zerowidthspace
> 
> * zerowidthjoiner
> 
> * zerowidthnonjoiner
> 
> * dottedCircle

Thanks! Both your assumptions are correct. I'll fix this on my end too.

**Comment by Hoolean** [comment]
Cheers @kalapi, thanks for the confirmation; all merged in now.

> Updated glyphs in the upright master:
> 
> ```
> ભ્ર bh_ra-gujarati
> લ્ર l_ra-gujarati
> ```

I also noticed that /b_ra-gujarati was changed in the upright masters instead of /bh_ra-gujarati, is that all good too? (apologies if feature code means that this is equivalent)

**Comment by kalapi** [comment]
> Cheers @kalapi, thanks for the confirmation; all merged in now.
> 
> > Updated glyphs in the upright master:
> > ```
> > ભ્ર bh_ra-gujarati
> > લ્ર l_ra-gujarati
> > ```
> 
> I also noticed that /b_ra-gujarati was changed in the upright masters instead of /bh_ra-gujarati, is that all good too? (apologies if feature code means that this is equivalent)

Sorry it's meant to be `બ્ર b_ra-gujarati`, which has indeed been fixed.

**Comment by Hoolean** [comment]
Super! We will plan the final QA 🚀

**Comment by MariannaPaszkowska** [comment]
Thank you all! This is our top priority for this week.

**Comment by Hoolean** [comment]
To update: Alex will be starting the final QA tomorrow morning; we will pass on any feedback as soon as possible :+1:

**Comment by chrissimpkins** [comment]
> To update: Alex will be starting the final QA tomorrow morning; we will pass on any feedback as soon as possible 👍

Tyvm for the update Harry!

**Comment by Hoolean** [comment]
@MariannaPaszkowska @chrissimpkins @kalapi

Hello,

Alex's QA has now concluded, with the latest feedback available [here](https://docs.google.com/document/d/1hy_BWuNmJWrbrZ4LOkq-DHvln0FRD35BDVnBEKYiCRc/edit?usp=sharing). In summary, most points have been addressed, although there is a potential issue with kerning, and with matra placement.

(The InDesign and /space issues documented are out-of-scope)

**Comment by MariannaPaszkowska** [comment]
Thank you, @Hoolean

**Comment by MariannaPaszkowska** [comment]
@Hoolean Could you point me to where that potential issue with kerning and the issue with matra placement are described?

**Comment by MariannaPaszkowska** [comment]
@kalapi Thanks for all your hard work so far. DaMa team is reporting that those issues are persisting in Beta 5: 

Screenshot 2022-10-13 at 2 41 29 PM

Screenshot 2022-10-13 at 2 41 55 PM

Screenshot 2022-10-13 at 2 42 06 PM

Screenshots were taken from [Google Sans Gujarati Import Log]( https://docs.google.com/document/d/1hy_BWuNmJWrbrZ4LOkq-DHvln0FRD35BDVnBEKYiCRc/edit# )

Could you please explain if this is intentional? 

cc @chrissimpkins

**Comment by Hoolean** [comment]
@MariannaPaszkowska @chrissimpkins Alex is on standby to absorb the QA tomorrow if we are able to merge new sources for then :+1:

**Comment by kalapi** [comment]
Hi @MariannaPaszkowska @Hoolean,

1. I can only reproduce the first issue with the GPOS which I've now fixed. 

- Fixes top anchors in `sa-gujarati` and `n_sa-gujarati` 

2. I cannot reproduce the kerning issue. It's fixed in my test fonts in BETA05, here is switching DIST on/off in FontGoggles:

DIST-ONOFF

3. The third issue is also a false positive, if you inspect my sources, you'll see that the space glyph has the correct value. 

Here is BETA06 with the GPOS fix:

[2022-10-13_GoogleSansGujarati_V1.000_BETA06.zip](https://github.com/googlefonts/googlesans/files/9777045/2022-10-13_GoogleSansGujarati_V1.000_BETA06.zip)

**Comment by Hoolean** [comment]
Thank you for the new sources @kalapi : ) I will pass them over to Alex, who will conclude the review, and we will aim to get the all-clear tomorrow.

---

> Fixes top anchors in `sa-gujarati` and `n_sa-gujarati`

I noticed in the adjustments that "bottom" and "topright" were removed; is this just from pruning unused anchors?

> I cannot reproduce the kerning issue. It's fixed in my test fonts in BETA05, here is switching DIST on/off in FontGoggles

Yes, this was our mistake, sorry about that; I will ensure we use DIST instead of our kerning toggle when checking these changes.

> The third issue is also a false positive, if you inspect my sources, you'll see that the space glyph has the correct value.

In the beta 6 sources/binaries, the space widths still vary across Display Italic; this may be where the confusion lies, as the values look correct elsewhere. Nevertheless, we do not need to correct this now as it is not imported, so we are safe to leave it as it is.

**Comment by kalapi** [comment]
Hello

> I noticed in the adjustments that "bottom" and "topright" were removed; is this just from pruning unused anchors?

Yes indeed. Those were artefacts from de-componentisation. 

> > The third issue is also a false positive, if you inspect my sources, you'll see that the space glyph has the correct value.
> 
> In the beta 6 sources/binaries, the space widths still vary across Display Italic; this may be where the confusion lies, as the values look correct elsewhere. Nevertheless, we do not need to correct this now as it is not imported, so we are safe to leave it as it is.

I've fixed this on our side. I figured out the confusion and have incorporated a check for this in future releases. 

Thanks for your patience :)

**Comment by MariannaPaszkowska** [comment]
Thank you for the speedy turnaround! @kalapi @Hoolean

**Comment by Hoolean** [comment]
@MariannaPaszkowska @chrissimpkins

Hello!

Alex [has completed the review](https://docs.google.com/document/d/1hy_BWuNmJWrbrZ4LOkq-DHvln0FRD35BDVnBEKYiCRc/edit#bookmark=id.mt7v6pvl2sq3), and all non-InDesign issues have been resolved :tada: Thank you @kalapi!

I have set a build branch running [here](https://github.com/googlefonts/googlesans/actions/runs/3249137456); after we have checked the CI warnings, we will be ready to merge 🚀

**Comment by Hoolean** [comment]
All CI checks passing; no additional warnings when compared to `main`.

**Review by MariannaPaszkowska — APPROVED** [review]

File: scripts/one-off/check-script-tags.py (hunk 1 above)
Comment (chrissimpkins) [reply scripts/one-off/check-script-tags.py:15]: Ty!

**Comment by chrissimpkins** [comment]
Many thanks to all for all of your hard work on this!

File: scripts/one-off/check-script-tags.py (hunk 1 above)
Comment (chrissimpkins) [reply scripts/one-off/check-script-tags.py:15]: https://github.com/googlefonts/googlesans/pull/534/commits/584f66c8249ab8e0adc8e82bfe73c2d7fd363846

**Review by chrissimpkins — APPROVED** [review]

**Comment by Hoolean** [comment]
I have rebased on `main`, and now the diffs in the shaping tests all stem from Gujarati. When regenerating the shaping JSONs, we see no changes also : )

I have set a build branch going, and will let you know when we get the all clear there :+1:

**Comment by Hoolean** [comment]
@MariannaPaszkowska @chrissimpkins

The CI [says all-clear](https://github.com/googlefonts/googlesans/actions/runs/3266777549) following the rebase: no warnings beyond those for Odia 🚀

**Comment by MariannaPaszkowska** [comment]
Thank you, @kalapi, @Hoolean, @madig, @belluzj, @chrissimpkins! 🎉

**Pull request merged**